### DISCUSSION TITLE
Does Talos support real-time kernel

### DISCUSSION THREAD

**Discussion by yingwana3** [siderolabs/talos#7441] [Q&A] [answered]
Hi, all

We need to run time-sensisive workload on Talos. Is there any document for changing default kernel to rt-kernel? Besides, how can we change kernel cmaline parameters after the cluster is installed. We try to modify the controlplane.yaml on extraKernelArgs part and re-apply the config, but talosctl cat /proc/cmdline shows no change at all.

Thanks

**Comment by smira** [comment, marked as answer]
Talos officially supports only a single Linux kernel it was tested with. You can build Talos with your kernel image and use it.

`extraKernelArgs` are part of the `install` section, so they are only applied atm when you do an `upgrade` (or fresh install), so workaround is to perform an upgrade to the same version of Talos, e.g. `talosctl upgrade -n <node> --preserve -i ghcr.io/siderolabs/installer:v1.4.6`

**Reply by yingwana3** [reply]
Thanks a lot for your quick reply.

**Reply by yumin-l** [reply]
We have installed talos iso on bare metal platform with https://github.com/siderolabs/talos/releases/download/v1.4.4/talos-amd64.iso
Then, we want to customize rt-kernel and build & install it with talos

We are trying to install talos with our own installer built with rt kernel, but we failed.
The steps used are as follows:
(Is this the right approach to change talos to rt kernel? Or what's the right approach to do so?)

- git clone https://github.com/talos-systems/pkgs.git
- Replace the url in pkg/kernel/prepare/pkg.yaml
![image](https://github.com/siderolabs/talos/assets/101164877/ceec99df-c480-4042-a4da-6c01e151fabb)

-  Replace the content in pkg/Pkgfile
![image](https://github.com/siderolabs/talos/assets/101164877/891baa70-846b-470b-9ecf-bf8d846f9dd9)

- make kernel-menuconfig
- make kernel
- DOCKER_BUILDKIT=0 docker build --build-arg RM="/lib/modules" -t installer:kernel .
- talosctl gen secrets -o secrets.yaml and talosctl gen config --with-secrets secrets.yaml <cluster-name> <cluster-endpoint>
- Replace the content in the generated file controlplane.yaml
![image](https://github.com/siderolabs/talos/assets/101164877/9681f9ff-09f8-4fb9-ad3f-a08f7b54cb14)

- talosctl apply-config

Everytime we apply the config above. The system will freeze and then reboot.
Does the iso version have anything to do with the image in the file controlplane.yaml?

**Reply by frezbo** [reply]
it seems `DOCKER_BUILDKIT=0 docker build --build-arg RM="/lib/modules" -t installer:kernel .` it;s missing `ghcr.io` prefix, was the image pushed to the registry?. I would help if the console logs are available.

I would also suggest to test this with local qemu cluster

`talosctl cluster create --provisioner=qemu --install-image=<your custom image>`

The logs would be in `~/.talos/cluster` for each cluster, would be easy to debug

**Reply by yingwana3** [reply]
hi, frezbo

Our platform is bare mental server with Ubuntu 2204 OS.

We use your official image to test the qemu cluser and  attach the error logs as below:

# talosctl cluster create --provisioner=qemu --install-image=ghcr.io/siderolabs/installer:v1.4.6
validating CIDR and reserving IPs
generating PKI and tokens
creating state directory in "/root/.talos/clusters/talos-default"
creating network talos-default
creating load balancer
creating dhcpd
creating controlplane nodes
creating worker nodes
renamed talosconfig context "talos-default" -> "talos-default-3"
waiting for API
bootstrap error: 3 error(s) occurred:
        rpc error: code = DeadlineExceeded desc = context deadline exceeded
        rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp 10.5.0.2:50000: connect: no route to host"
        timeout


Besides, we also meet an issue with talos quickstart following this link: https://www.talos.dev/v1.4/introduction/quickstart/
# talosctl cluster create
validating CIDR and reserving IPs
generating PKI and tokens
creating network talos-default
creating controlplane nodes
creating worker nodes
renamed talosconfig context "talos-default" -> "talos-default-2"
waiting for API
bootstrapping cluster
◳ waiting for etcd to be healthy: 1 error occurred:
    * 10.5.0.2: service "etcd" not in expected state "Running": current state [Preparing] Running pre state

We have closed the firewall and set $no_proxy for our local IP and 10.5.0.2.

Hope you can give some hint.

Thanks

**Reply by frezbo** [reply]
you can check the dmesg from files in `~/.talos/cluster/<cluster-name>/*.log` this doesn't provide much info

**Reply by frezbo** [reply]
`creating state directory in "/root/.talos/clusters/talos-default"` well this path specifically

**Reply by yingwana3** [reply]
Hi, attach thie log info, please help to have a look.

cat dhcpd.log
[dhcpv4] 2023/07/17 00:10:58 Server listening on 0.0.0.0:67
[dhcpv4] 2023/07/17 00:10:58 Ready to handle requests

cat lb.log
2023/07/17 00:10:58 healthcheck failed for "10.5.0.2:6443": dial tcp 10.5.0.2:6443: i/o timeout
2023/07/17 00:10:59 healthcheck failed for "10.5.0.2:6443": dial tcp 10.5.0.2:6443: i/o timeout
2023/07/17 00:11:00 healthcheck failed for "10.5.0.2:6443": dial tcp 10.5.0.2:6443: i/o timeout
2023/07/17 00:11:01 healthcheck failed for "10.5.0.2:6443": dial tcp 10.5.0.2:6443: i/o timeout
2023/07/17 00:11:02 healthcheck failed for "10.5.0.2:6443": dial tcp 10.5.0.2:6443: connect: no route to host

cat talos-default-controlplane-1.log
starting /usr/bin/qemu-system-x86_64 with args:
-m 2048 -smp cpus=2 -cpu max -nographic -netdev tap,id=net0,ifname=tap0,script=no,downscript=no -device virtio-net-pci,netdev=net0,mac=42:20:90:6b:6b:33 -device virtio-rng-pci -device virtio-balloon,deflate-on-oom=on -monitor unix:/root/.talos/clusters/talos-default/talos-default-controlplane-1.monitor,server,nowait -no-reboot -boot order=cn,reboot-timeout=5000 -smbios type=1,uuid=8429c8e3-e53f-4048-84bd-c6f5d1fbaca7 -drive format=raw,if=virtio,file=/root/.talos/clusters/talos-default/talos-default-controlplane-1-0.disk,cache=unsafe -machine q35,accel=kvm -drive file=/root/.talos/clusters/talos-default/talos-default-controlplane-1-flash0.img,format=raw,if=pflash -kernel _out/vmlinuz-amd64 -initrd _out/initramfs-amd64.xz -append init_on_alloc=1 slab_nomerge pti=on consoleblank=0 nvme_core.io_timeout=4294967295 printk.devkmsg=on ima_template=ima-ng ima_appraise=fix ima_hash=sha512 console=ttyS0 reboot=k panic=1 talos.shutdown=halt talos.platform=metal talos.config=http://10.5.0.1:35669/config.yaml
qemu: could not open kernel file '_out/vmlinuz-amd64': No such file or directory
process exited with error exit status 1

cat talos-default-worker-1.log
starting /usr/bin/qemu-system-x86_64 with args:
-m 2048 -smp cpus=2 -cpu max -nographic -netdev tap,id=net0,ifname=tap0,script=no,downscript=no -device virtio-net-pci,netdev=net0,mac=8a:0e:62:0d:f1:74 -device virtio-rng-pci -device virtio-balloon,deflate-on-oom=on -monitor unix:/root/.talos/clusters/talos-default/talos-default-worker-1.monitor,server,nowait -no-reboot -boot order=cn,reboot-timeout=5000 -smbios type=1,uuid=4a195349-7561-43c3-8270-34bd08775c3e -drive format=raw,if=virtio,file=/root/.talos/clusters/talos-default/talos-default-worker-1-0.disk,cache=unsafe -machine q35,accel=kvm -drive file=/root/.talos/clusters/talos-default/talos-default-worker-1-flash0.img,format=raw,if=pflash -kernel _out/vmlinuz-amd64 -initrd _out/initramfs-amd64.xz -append init_on_alloc=1 slab_nomerge pti=on consoleblank=0 nvme_core.io_timeout=4294967295 printk.devkmsg=on ima_template=ima-ng ima_appraise=fix ima_hash=sha512 console=ttyS0 reboot=k panic=1 talos.shutdown=halt talos.platform=metal talos.config=http://10.5.0.1:35477/config.yaml
qemu: could not open kernel file '_out/vmlinuz-amd64': No such file or directory
process exited with error exit status 1

**Reply by frezbo** [reply]
as the error says you don;t have the kernel file:

`qemu: could not open kernel file '_out/vmlinuz-amd64': No such file or directory`

you'd have to either copy kernel and initramfs to `_out` or run `make kernel initramfs` which would copy the file.

**Reply by yingwana3** [reply]
Hi, frezo

As we are totally new to talos. Whether it's possible to provide some guidance on how to build a talos-rt-kernel.iso (we can customize the realtime kernel version). Currently, we can only install talos successfully with this way.

**Reply by frezbo** [reply]
if you have the custom kernel already built you can place that under `_out/vmlinuz-amd64` and run `make initramfs` to create the initramfs and test
